tests: sprintf: zero-length gnu_printf format string

**Describe the bug**
The print function is called and the function is not intended to be used to print characters, in this case shows build failure.

For example :
tests/lib/sprintf/src/main.c:778:24    ret = fprintf(stdout, "");
usb/device/usb_dc_kinetis.c:840:10     LOG_DBG("");
tests/lib/sprintf/src/main.c:760:15    ret = printf("");

It seems to be an error to print noting with twister command , but it is fine to use west build command.

Take tests/lib/sprintf/ as an example .

**To Reproduce**
Steps to reproduce the behavior:
1. source zephyr-env.sh
2. twister -p frdm_k64f --device-testing --hardware-map map.yml -T tests/lib/sprintf/ -vv
3. See error

**Expected behavior**
The test passes.

**Impact**
The test can not run by twister command.

**Logs and console output**
```
[25/136] Building C object CMakeFiles/app.dir/src/main.c.obj
FAILED: CMakeFiles/app.dir/src/main.c.obj
ccache /home/ztest/zephyr-sdk-0.14.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DBOARD_FLASH_SIZE="CONFIG_FLASH_SIZE*1024" -DCPU_MK64FN1M0VLL12 -DKERNEL -DTC_RUNID=ec44835519e9be9f92aef986773c3ea9 -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR__=1 -I/home/ztest/zephyrproject/zephyr/include/zephyr -I/home/ztest/zephyrproject/zephyr/include -I/home/ztest/zephyrproject/zephyr/twister-out/frdm_k64f/tests/lib/sprintf/libraries.libc.sprintf/zephyr/include/generated -I/home/ztest/zephyrproject/zephyr/soc/arm/nxp_kinetis/k6x -I/home/ztest/zephyrproject/zephyr/soc/arm/nxp_kinetis/common -I/home/ztest/zephyrproject/zephyr/subsys/testsuite/include -I/home/ztest/zephyrproject/zephyr/subsys/testsuite/ztest/include -I/home/ztest/zephyrproject/zephyr/subsys/testsuite/include/zephyr -I/home/ztest/zephyrproject/zephyr/subsys/testsuite/ztest/include/zephyr -I/home/ztest/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include -I/home/ztest/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/devices/MK64F12 -I/home/ztest/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/devices/MK64F12/drivers -I/home/ztest/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/drivers/common -I/home/ztest/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/drivers/uart -I/home/ztest/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/drivers/port -I/home/ztest/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/drivers/sysmpu -isystem /home/ztest/zephyrproject/zephyr/lib/libc/minimal/include -isystem /home/ztest/zephyr-sdk-0.14.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/include -isystem /home/ztest/zephyr-sdk-0.14.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/include-fixed -Os -imacros /home/ztest/zephyrproject/zephyr/twister-out/frdm_k64f/tests/lib/sprintf/libraries.libc.sprintf/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfpu=fpv4-sp-d16 -mfloat-abi=hard -mfp16-format=ieee --sysroot=/home/ztest/zephyr-sdk-0.14.1/arm-zephyr-eabi/arm-zephyr-eabi -imacros /home/ztest/zephyrproject/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -Werror -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/ztest/zephyrproject/zephyr/tests/lib/sprintf=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/ztest/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/ztest/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT CMakeFiles/app.dir/src/main.c.obj -MF CMakeFiles/app.dir/src/main.c.obj.d -o CMakeFiles/app.dir/src/main.c.obj -c /home/ztest/zephyrproject/zephyr/tests/lib/sprintf/src/main.c
/home/ztest/zephyrproject/zephyr/tests/lib/sprintf/src/main.c: In function 'test_print':
/home/ztest/zephyrproject/zephyr/tests/lib/sprintf/src/main.c:760:15: error: zero-length gnu_printf format string [-Werror=format-zero-length]
  760 |  ret = printf("");
      |               ^~
/home/ztest/zephyrproject/zephyr/tests/lib/sprintf/src/main.c: In function 'test_fprintf':
/home/ztest/zephyrproject/zephyr/tests/lib/sprintf/src/main.c:778:24: error: zero-length gnu_printf format string [-Werror=format-zero-length]
  778 |  ret = fprintf(stdout, "");
      |                        ^~
cc1: all warnings being treated as errors
[26/136] Building C object zephyr/lib/posix/CMakeFiles/lib__posix.dir/pthread_common.c.obj
[27/136] Building C object zephyr/lib/libc/minimal/CMakeFiles/lib__libc__minimal.dir/source/stdlib/abort.c.obj
[28/136] Building C object zephyr/lib/libc/minimal/CMakeFiles/lib__libc__minimal.dir/source/stdlib/strtoul.c.obj
[29/136] Building C object zephyr/lib/libc/minimal/CMakeFiles/lib__libc__minimal.dir/source/stdlib/strtoll.c.obj
[30/136] Building C object zephyr/lib/libc/minimal/CMakeFiles/lib__libc__minimal.dir/source/stdlib/strtoull.c.obj
[31/136] Building C object zephyr/lib/posix/CMakeFiles/lib__posix.dir/nanosleep.c.obj
[32/136] Building C object zephyr/arch/arch/arm/core/aarch32/mpu/CMakeFiles/arch__arm__core__aarch32__mpu.dir/nxp_mpu.c.obj
[33/136] Building C object zephyr/subsys/testsuite/ztest/CMakeFiles/subsys__testsuite__ztest.dir/src/ztest.c.obj
[34/136] Building C object zephyr/arch/arch/arm/core/aarch32/cortex_m/CMakeFiles/arch__arm__core__aarch32__cortex_m.dir/fault.c.obj
ninja: build stopped: subcommand failed.

```

**Environment :**
OS: Linux,
Toolchain Zephyr SDK 0.14.1
Commit SHA or Version: b161554ee862

**Additional context**
The same error here: 
tests/lib/sprintf/src/main.c:760:15                  ret = printf("");
tests/lib/sprintf/src/main.c:778:24                  ret = fprintf(stdout, "");
drivers/usb/device/usb_dc_kinetis.c:188:10    LOG_DBG("");
drivers/usb/device/usb_dc_kinetis.c:188:2      LOG_DBG("");
drivers/usb/device/usb_dc_kinetis.c:221:10    LOG_DBG("");
drivers/usb/device/usb_dc_kinetis.c:251:10    LOG_DBG("");
drivers/usb/device/usb_dc_kinetis.c:261:10    LOG_DBG("");
drivers/usb/device/usb_dc_kinetis.c:810:10    LOG_DBG("");
drivers/usb/device/usb_dc_kinetis.c:840:10    LOG_DBG("");
subsys/usb/device/class/hid/core.c:617:10     LOG_DBG("");
subsys/usb/device/class/hid/core.c:617:2       LOG_DBG("");

--- a/cmake/compiler/gcc/compiler_flags.cmake
+++ b/cmake/compiler/gcc/compiler_flags.cmake
@@ -28,9 +28,8 @@ set_compiler_property(PROPERTY optimization_size  -Os)
 # GCC Option standard warning base in Zephyr
 check_set_compiler_property(PROPERTY warning_base
     -Wall
-    -Wformat
-    -Wformat-security
-    -Wno-format-zero-length
+    "SHELL:-Wformat -Wformat-security"
+    "SHELL:-Wformat -Wno-format-zero-length"
     -Wno-main
 )
 

--- a/cmake/modules/extensions.cmake
+++ b/cmake/modules/extensions.cmake
@@ -1048,7 +1048,7 @@ function(zephyr_check_compiler_flag lang option check)
   set(key_string "${key_string}${option}_")
   set(key_string "${key_string}${CMAKE_REQUIRED_FLAGS}_")
 
-  string(MD5 key ${key_string})
+  string(MD5 key "${key_string}")
 
   # Check the cache
   set(key_path ${ZEPHYR_TOOLCHAIN_CAPABILITY_CACHE_DIR}/${key})
@@ -1068,7 +1068,7 @@ function(zephyr_check_compiler_flag lang option check)
   # tested, so to test -Wno-<warning> flags we test -W<warning>
   # instead.
   if("${option}" MATCHES "-Wno-(.*)")
-    set(possibly_translated_option -W${CMAKE_MATCH_1})
+    string(REPLACE "-Wno-" "-W" possibly_translated_option "${option}")
   else()
     set(possibly_translated_option ${option})
   endif()
@@ -1814,7 +1814,7 @@ function(check_compiler_flag lang option ok)
   endif()
 
   string(MAKE_C_IDENTIFIER
-    check${option}_${lang}_${CMAKE_REQUIRED_FLAGS}
+    "check${option}_${lang}_${CMAKE_REQUIRED_FLAGS}"
     ${ok}
     )
 
@@ -2004,6 +2004,12 @@ endfunction()
 # with the extension that it will check that the compiler supports the flag
 # before setting the property on compiler or compiler-cpp targets.
 #
+# To test flags together, such as '-Wformat -Wformat-security', an option group
+# can be specified by using shell-like quoting along with a 'SHELL:' prefix.
+# The 'SHELL:' prefix will be dropped before testing, so that
+# '"SHELL:-Wformat -Wformat-security"' becomes '-Wformat -Wformat-security' for
+# testing.
+#
 # APPEND: Flag indicated that the property should be appended to the existing
 #         value list for the property.
 # PROPERTY: Name of property with the value(s) following immediately after
@@ -2021,8 +2027,13 @@ function(check_set_compiler_property)
   list(REMOVE_AT COMPILER_PROPERTY_PROPERTY 0)
 
   foreach(option ${COMPILER_PROPERTY_PROPERTY})
+    if(${option} MATCHES "^SHELL:")
+      string(REGEX REPLACE "^SHELL:" "" option ${option})
+      separate_arguments(option UNIX_COMMAND ${option})
+    endif()
+
     if(CONFIG_CPLUSPLUS)
-      zephyr_check_compiler_flag(CXX ${option} check)
+      zephyr_check_compiler_flag(CXX "${option}" check)
 
       if(${check})
         set_property(TARGET compiler-cpp ${APPEND-CPP} PROPERTY ${property} ${option})
@@ -2030,7 +2041,7 @@ function(check_set_compiler_property)
       endif()
     endif()
 
-    zephyr_check_compiler_flag(C ${option} check)
+    zephyr_check_compiler_flag(C "${option}" check)
 
     if(${check})
       set_property(TARGET compiler ${APPEND} PROPERTY ${property} ${option})

